[ongzhili] iP

ongzhili opened the pull request:
# Lamball Chatbot 🐐 
> Not to be confused with _Wooloo_, nope, not similar at all!

Lamball is a helpful pal that you never knew you needed. 

- [x] Helps you manage your tasks and deadlines
- [ ] Reminds you about them (coming soon!)
- [x] Guides you through the commands!

All you need to do is:

1. Find the latest release [here](https://github.com/ongzhili/ip/releases)
2. Double click on the file.
3. Add your tasks.

And it is 

- **FREE!**
- runs in 1 `main` command!
```java
    public static void main(String[] args) {
        new Lamball().run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

ongzhili committed:
Gave chatbot a new name
Skeletal Message added

ongzhili committed:
Improve the skeletal version of Duke so that it echos commands entered by the user, and exits when the user types the command bye.

ongzhili committed:
Real stage 1 commit (Misinterpreted instruction)

ongzhili committed:
Improved input parse handling

ongzhili committed:
Add, List functionalities added

ongzhili committed:
Merge remote-tracking branch 'origin/add-gradle-support' into HEAD

ongzhili committed:
Modified welcome Message
Cleared Level 3 and A-classes requirement

ongzhili committed:
Level 4 commits: Created subclasses based on provided format. Replaced task array to an ArrayList

ongzhili committed:
Basic text UI testing (A-TextUiTesting)

ongzhili committed:
1. Updated project to satisfy Level-5 requirements
2. Updated TextUiTesting accordingly to test new functionality

ongzhili committed:
1. Added Delete functionality (Level-6)
2. Added test cases accordingly

ongzhili committed:
Add some plaintext file reading functionality
1. Create plaintext file if does not exist
2. Read plaintext file contents

ongzhili committed:
Update .gitignore
1. Ignores saved list

ongzhili committed:
Add filter for invalid lines in text file
Current implementation is just to ignore invalid inputs

However, in the longer run, this might be difficult to handle as
more file operations are added.

Remove lines by overwriting current list file with another file
that only has the valid lines written.

ongzhili committed:
Reflect change in todo list in text file

Text file does not reflect change in internal list of todo tasks.

If the change is not reflected, progress is not correctly saved
across instances.

Add some file modifying methods to facilitate functionalities such
as:
1. Add Task
2. Delete Task (given index n)

ongzhili committed:
Merge branch 'branch-level-7'

ongzhili committed:
Implement Date functionality into Deadline and Event

Dates are just strings.

So, users could input deadlines that made no sense (i.e I have a
homework deadline by apple).

Restrict date format to yyyy-mm-dd to not only facilitate logical
dates in the list, but to open up to future date comparison features
as well.

ongzhili committed:
Separate methods into different classes based on functionality

Everything is in one class (except tasks which was done before)

Divide methods into classes based on their usage (i.e Storage
methods goes to Storage class, etc)

Code is more modular and clean this way

ongzhili committed:
Fix indent and format errors

Some status messages show extra indent and line.

User experience would be improved with consistent formatting

ongzhili committed:
Merge branch 'branch-a-oop'

ongzhili committed:
Package classes based on their uses

All classes are in the same directory.

Looks very disorganized. Separating them into different packages
improves project structure.

Move everything into a Lamball Package, then moved tasks into Tasks
package.

ongzhili committed:
Package classes based on their uses part 2

Classes in original directory not there anymore (moved into package)

Reflect change in directory and build.gradle

ongzhili committed:
Reformat parsing

Parse function did more than parsing

It is doing other classes' job for them, which makes for non-elegant
code

Clean up parse and move it's extra functionalities to where they
should belong

ongzhili committed:
Merge branch 'branch-a-betterer-oop'

ongzhili committed:
Reformat directories

L in Lamball capitalized in directory

Gradle has issues picking up JUnit tests

Reformat directories, changed Lamball to lamball (directory)
Create some sanity tests to verify gradle file configured properly

ongzhili committed:
Add JUnit testcases

Text Comparison is used in testing.

Text comparison between actual and expected does not test for things
like exceptions thrown

Use JUnit to test for non-trivial scenarios like exceptions thrown.

ongzhili committed:
Merge branch 'branch-a-junit'

ongzhili committed:
Format gradle build file

.jar file is named duke for a chatbot named Lamball

Misleading executable file name

Change shadowJar parameters to match chatbot name

ongzhili committed:
Merge branch 'branch-a-jar'

ongzhili committed:
Add JavaDoc to public methods

No documentation provided for methods.

Add descriptive JavaDoc documentation to better explain the code.

Providing JavaDoc documentation for public methods improves
readability and understandability of code when read by external
parties.

ongzhili committed:
Tweak code to comply with coding standard

Inconsistent coding standard used in code.

Desire to follow a standardized coding standard so that the code is
more readable.

Added some line breaks and adjusted certain variable naming to suit
the standard coding convention that was required.

ongzhili committed:
Tweak code to comply with coding standard p2

Deadline class still had methods that did not follow coding standard

Reflects on lack of due diligence if coding standard is found to not
be followed in numerous instances

Fix line break to separate method signatures.

This makes the Deadline class follow the coding standard set.

ongzhili committed:
Add find function in chatbot commands

Users are unable to search their todo lists with keywords or
markers.

This is bad user design, and especially when more tasks are added
to the todo list, a search function is a necessary feature for users
to filter.

Let's
*Add a 'find' function that is parsable by the parse.
*Update tasklist and other functions to faciliate integration of
said function.

ongzhili committed:
Merge branch 'branch-A-CodingStandard'

ongzhili committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/lamball/Storage.java
#	src/main/java/lamball/TaskList.java
#	src/main/java/lamball/Ui.java
#	src/main/java/lamball/task/Task.java

ongzhili committed:
Add header descriptions to classes

Header descriptions not present.

Bad for other readers and code reviewers because they would require
more time to understand the code and how each part functions
separately.

Add a small javadoc description for every class.

ongzhili committed:
Merge branch 'branch-A-JavaDoc'

ongzhili committed:
Move additional String printing related methods to UI

Printlist is printed in TaskList.

Bad because in the structure, Ui is supposed to be doing the Ui
work, and not TaskList.

Streamline all task commands to better allow Ui to do the UI
operations.

darkv32 commented on src/main/java/lamball/Lamball.java:
```diff
@@ -0,0 +1,77 @@
+        this.initialize();
+
+        while (isActive) {
+            System.out.print("    You:");
```
Although single line of code, should all UI related changes be moved to UI class?

darkv32 commented on src/main/java/lamball/Parser.java:
```diff
@@ -0,0 +1,53 @@
+     */
+    public static String[] parse(String msg) throws LamballParseException {
+        String[] parts = msg.split(" ", 2);
+        switch (parts[0]) {
```
Nice switch statement, cute exception messages

darkv32 commented on src/main/java/lamball/Storage.java:
```diff
@@ -0,0 +1,186 @@
+ */
+public class Storage {
+
+    protected String filePath = "src/main/java/data/list.txt";
+    protected String tempFilePath = "src/main/java/data/tempfile.txt";
```
Should these variables be final?

darkv32 commented on src/main/java/lamball/Storage.java:
```diff
@@ -0,0 +1,186 @@
+     */
+
+    public static void obtainSavedFile(Lamball lamb) {
+        File folder = new File("src/main/java/data");
```
Alot of reused path names, can you extract them out and append specific paths to it? i.e. `protected final String ROOT_PATH = "src/main/java/data";`

darkv32 commented on src/main/java/lamball/Storage.java:
```diff
@@ -0,0 +1,186 @@
+
+                // System.out.println("Line " + index + " replaced successfully.");
+            } else {
+                // System.err.println("aaaaaaa");
```
Empty else statement?

darkv32 commented on src/main/java/lamball/TaskList.java:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    private void printList(ArrayList<Task> lst) {
+        String listStr = "    Here aaaaare the taaaasks in your list:";
```
Method named print, but no printing here? Is `getList()`, or `updateList()` a better name?

darkv32 commented on src/main/java/lamball/TaskList.java:
```diff
@@ -0,0 +1,198 @@
+    private final String INDENT = "    ____________________________________________________________\n";
+    protected ArrayList<Task> tasks;
+
+    String lastDoneTask;
```
Weird name for the areas where its used.

darkv32 commented on src/main/java/lamball/TaskList.java:
```diff
@@ -0,0 +1,198 @@
+
+        }
+        // Should not reach here.
+        return false;
```
Can we move this unreachable to `default:` to catch unexpected commands?

darkv32 commented on src/main/java/lamball/Ui.java:
```diff
@@ -0,0 +1,62 @@
+ */
+
+public class Ui {
+    private static final String WELCOME_MESSAGE = "    lamball\n" +
```
Can we use `INDENT` to make this neater?

darkv32 commented on src/test/java/lamball/ParserTest.java:
```diff
@@ -0,0 +1,71 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class ParserTest {
```
Good variety of tests

darkv32 commented on src/test/java/lamball/task/DeadlineTest.java:
```diff
@@ -0,0 +1,36 @@
+import java.time.format.DateTimeParseException;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+public class DeadlineTest {
```
Can we split these tests? In case 1 fails, the rest can still run.

darkv32 commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,23 @@
+todo borrow book
```
Good variety of inputs

darkv32 commented on src/test/java/lamball/task/EventTest.java:
```diff
@@ -0,0 +1,38 @@
+public class EventTest {
+    @Test
+    public void eventTest() {
+        // Command Test
```
Can we split these tests? In case 1 fails, the rest can still run.

darkv32 commented on src/test/java/lamball/task/TaskTest.java:
```diff
@@ -0,0 +1,17 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+public class TaskTest {
+    @Test
+    public void taskTest() {
```
Can we split these tests? In case 1 fails, the rest can still run.

darkv32 commented on src/test/java/lamball/task/ToDoTest.java:
```diff
@@ -0,0 +1,18 @@
+public class ToDoTest {
+
+    @Test
+    public void todoTest() {
```
Can we split these tests? In case 1 fails, the rest can still run.

darkv32 reviewed (commented):
Relevant parts of code are in appropriate classes, making it easy to read and track.
Some minor violations were noted, but nothing major.
Most classes were tested which is very good.
Test cases have to be split into different methods, so that if 1 test case fails early on, the rest can still run.

ongzhili committed:
Update gradle build to integrate javafx implementation

Build file does not contain dependencies for javafx.

Program cannot run without these dependencies.

Update dependencies.

ongzhili committed:
Implement GUI for chatbot

Running .jar file in terminal not very user-friendly.

A GUI might be a better option for users to use the chatbot.

Implement basic GUI with same output as original text.

ongzhili committed:
Merge branch 'branch-level-10'

ongzhili committed:
Update README.md

ongzhili committed:
Amend code to fit style guidelines

Many style violations present in code.

Not following style guide makes code untidy and sometimes unpleasant
to look at.

Used checkstyle script to check and amend style violations.

ongzhili committed:
Update README

Readme has Duke template readme.

Might be confusing to readers who want to view my code.

A new readme should be made for my new product.

Updated readme with gfmd formatting.

J-wang-CSMA commented on src/main/java/lamball/Lamball.java:
```diff
@@ -0,0 +1,222 @@
+ *
+ * @author ongzhili
+ */
+public class Lamball extends Application {
```
Perhaps you can consider placing the GUI implementation into the ui package while letting the Lamball package remain as the API? Or is there a specific reason why they are together for this application? 🤔

J-wang-CSMA commented on src/main/java/lamball/Lamball.java:
```diff
@@ -0,0 +1,222 @@
+import javafx.stage.Stage;
+import lamball.ui.DialogBox;
+
+/**
```
Like the Javadocs! Makes the program easy to read and understand.

J-wang-CSMA commented on src/main/java/lamball/Lamball.java:
```diff
@@ -0,0 +1,222 @@
+            String userInput = scanner.nextLine();
+
+            // Echo the user's command
+            try {
```
Good job on the error handling to prevent runtime crashes from bad inputs!

J-wang-CSMA commented on src/main/java/lamball/Parser.java:
```diff
@@ -0,0 +1,53 @@
+        case "deadline":
+        case "event": {
+            if (parts.length < 2 || parts[1] == null || parts[1].trim().isEmpty()) {
+                // Checks if empty string (nothing after command) or only whitespaces
```
I like the commenting, helps me to read through the code :)

J-wang-CSMA commented on src/main/java/lamball/LamballParseException.java:
```diff
@@ -0,0 +1,17 @@
+ *
+ * @author ongzhili
+ */
+public class LamballParseException extends Exception {
```
Interesting, are all exceptions specific to the Lamball application due to parsing issues, since the name of the error seems to suggest so?

J-wang-CSMA commented on src/main/java/lamball/Storage.java:
```diff
@@ -0,0 +1,193 @@
+        }
+    }
+
+    private static void initializeListFromText(File savedList, Lamball lamb) throws FileNotFoundException {
```
Perhaps the naming of the methods can be more consistent here? We have:
1. Load = obtainSavedFile
2. Save = writeToFile (There's method overloading as well, might want to consider commenting or not overload if the methods are doing different things.)

J-wang-CSMA commented on src/main/java/lamball/TaskList.java:
```diff
@@ -0,0 +1,207 @@
+     * @return Boolean indicating whether to keep the chatbot active after the command.
+     * @throws LamballParseException if invalid arguments provided.
+     */
+    public boolean runComd(String[] command, boolean isInit) throws LamballParseException {
```
Should the logic for running the commands be in the TaskList class?

J-wang-CSMA commented on src/main/java/lamball/TaskList.java:
```diff
@@ -0,0 +1,207 @@
+        return this.lastDoneTask;
+    }
+
+    private void printList(ArrayList<Task> lst) {
```
Perhaps we can spell out the argument name fully, lst can look like 1st to some. :)

J-wang-CSMA commented on src/main/java/lamball/Ui.java:
```diff
@@ -0,0 +1,68 @@
+ */
+
+public class Ui {
+    private static final String WELCOME_MESSAGE =
```
Good usage of constants!

J-wang-CSMA commented on src/main/java/lamball/task/Task.java:
```diff
@@ -0,0 +1,62 @@
+     * @param compare String used for comparison.
+     * @return Whether the given string is a sub-sequence of the task.
+     */
+    public boolean containing(String compare) {
```
Maybe you might want to consider naming this in a more informative way (and in a way that sounds more like a boolean method), such as containsKeyword?

J-wang-CSMA reviewed (commented):
Looks good to me overall. Might want to consider applying some more OOP principles to the code, for instance, to encapsulate the handling of different commands within a Command class perhaps? 

Could also make use of packages a bit more. For instance, maybe you would like to consider putting the UI implementation into a separate package that can then be easily modified if we want to change something about the UI without breaking the other code?

Saw the pictures and the personality of the bot, it is very cute, good job!

ongzhili committed:
Update classes based on PR review

Recieved comments that more OOP principles could be used in various
aspects of my code.

Code is messy and violates some OOP principles if not changed.

Add command package to handle commands.

ongzhili committed:
Update tests, fix checkstyle

Code not tidy after big overhaul of old code and addition of new
code.

It is crucial to maintain good code quality to ensure readability.

Update tests and add JavaDoc for new methods and classes, fix
coding style violations using script.

ongzhili committed:
Merge branch 'branch-a-peer-feedback'

ongzhili committed:
Improve code quality

Many magic literals currently present, some methods are extractable.

These might make the code confusing for readers who are not familiar
with how the code works.

Replace most of the literals with final variables and grouped some
code fragments together into a separate method.

ongzhili committed:
Add some key assertions

ongzhili committed:
Add assertion fail messages

ongzhili committed:
Remove obvious comments, unused commented code

ongzhili committed:
Merge pull request #2 from ongzhili/branch-a-codequality

Improve code quality

ongzhili committed:
Merge branch 'master' into branch-a-assertions

# Conflicts:
#	src/main/java/lamball/Storage.java

ongzhili committed:
Merge pull request #1 from ongzhili/branch-a-assertions

Add assertions

ongzhili committed:
Integrate CI into project

ongzhili committed:
Merge branch 'branch-a-ci'

ongzhili committed:
Update CI file

ongzhili committed:
Update CI file 2

ongzhili committed:
Update CI file 3

ongzhili committed:
Update CI file 4

ongzhili committed:
Update gitignore

ongzhili committed:
Add memo functionality

ongzhili committed:
Merge branch 'branch-d-notes'

ongzhili committed:
Adjust classes based on Jeremy's feedback

Jeremy commented that the UI was in my chatbot class, and that it
would be better if I were to move it to a separate class entirely.

I agree with the statement, as UI and chatbot functions should be
clearly separated.

Create new MainApp class to house GUI.

ongzhili committed:
Merge branch 'branch-a-peer-feedback-2'

ongzhili committed:
Add more error handling

ongzhili committed:
Merge branch 'branch-a-moreerrorhandling'

ongzhili committed:
Add some font adjustments

ongzhili committed:
Merge branch 'branch-a-bettergui'

ongzhili committed:
Add Ui image, remove unused variables

ongzhili committed:
Fix SoC SE Script feedback

ongzhili committed:
Merge branch 'branch-a-bettergui'

# Conflicts:
#	src/main/java/lamball/Lamball.java

ongzhili committed:
Rename Ui.PNG to Ui.png

ongzhili committed:
Update README.md

ongzhili committed:
Fix formatting errors

ongzhili committed:
Update README.md

ongzhili committed:
Fix table not rendering in github pages

ongzhili committed:
Fix image not rendering

ongzhili committed:
Update README.md

ongzhili committed:
Update README.md

ongzhili committed:
Fix minor spelling mistake

ongzhili committed:
Format example outputs

ongzhili committed:
Fix memo list index

ongzhili committed:
Merge branch 'master' of https://github.com/ongzhili/ip

ongzhili committed:
Update README.md
